### PR TITLE
Channel exception handler

### PR DESCRIPTION
**Pull request by siy**
Fix for https://radixdlt.atlassian.net/browse/NT-307

### CODE REVIEW

File: radixdlt-regression/build.gradle [hunk 1]
```diff
@@ -121,7 +121,7 @@ dependencies {
     api project(':radixdlt-java')
     api project(':radixdlt')
 
-    api 'com.github.bitcoinj:bitcoinj:7c31dcb'
+    api 'org.bitcoinj:bitcoinj-core:0.15.10'
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-regression/build.gradle:124]: This dependency is no longer available.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/transport/ExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  @Override
+  public void connect(
+      ChannelHandlerContext ctx,
+      SocketAddress remoteAddress,
+      SocketAddress localAddress,
+      ChannelPromise promise) {
+    ctx.connect(
+        remoteAddress, localAddress, promise.addListener(future -> handleFailure(future, ctx)));
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/transport/ExceptionHandler.java:90]: `exceptionCaught` is invoked only for "inbound" events, like `read` or `accept`. For "outbound" events, exceptions should be handled like I did above. There are two options - set two exception handlers (one at the beginning of pipeline and one at the end of pipeline) or use a single "universal" exception handler like one I've used. Full discussion could be found here https://stackoverflow.com/questions/30994095/how-to-catch-all-exception-in-netty

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/transport/PeerChannelInitializer.java [hunk 3]
```diff
@@ -162,7 +162,8 @@ private void createProxyProtocolPipeline(SocketChannel socketChannel) {
         .pipeline()
         .addLast("decode_proxy_header_line", new LineBasedFrameDecoder(255, true, true))
         .addLast("decode_proxy_header_bytes", new ByteArrayDecoder())
-        .addLast("handle_proxy_header", new ProxyHeaderHandler(socketChannel));
+        .addLast("handle_proxy_header", new ProxyHeaderHandler(socketChannel))
+        .addLast("exception_handler", new ExceptionHandler());
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/transport/PeerChannelInitializer.java:166]: Good catch. Fixed.

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/transport/ExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  private static void handleFailure(Future<?> future, ChannelHandlerContext ctx) {
+    if (!future.isSuccess()) {
+      logAndCloseChannel(ctx, future.cause());
+    }
+  }
+
+  private static void logAndCloseChannel(ChannelHandlerContext ctx, Throwable cause) {
+    log.warn("Exception caught in channel, closing channel to prevent resource leak", cause);
```
[the remark this answers is not in the document]
Comment (siy) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/transport/ExceptionHandler.java:105]: Fixed

**Review by LukasGasior1 — APPROVED** [review]
LGTM

File: radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/transport/ExceptionHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhedey) [reply radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/transport/ExceptionHandler.java:97]: Thanks Sergiy! Could we add a link to this documentation in a comment, if it's potentially a little unexpected? :)

**Pull request merged**